backend/app/consents.py

```diff
@@ -0,0 +1,184 @@
+        headers=headers,
+    )
+    return token
+
+
+def make_dsi_uri(definition: str) -> str:
+    """
+    Make Data Source Identifier
+    :param definition: Data Product Definition
+    :return: Data Source Identifier
+    """
+    return f"dpp://{conf.DATA_PRODUCT_SOURCE}@{conf.DATASPACE_DOMAIN}/{definition}"
+
+
+async def request_consent_token(dsi: str, sub: str) -> Optional[str]:
+    """
+    Fetch Consent Token from Consent Portal
+    :param dsi: Data Source Identifier
+    :param sub: Subject for a consent
+    :return: Consent Token if it has been granted already, None otherwise
+    """
+    url = f"{conf.CONSENT_PORTAL_URL}/Consent/GetToken"
+    cr_token = create_consent_request_token(sub)
+    headers = {"X-Consent-Request-Token": cr_token}
+    payload = {"dataSource": dsi}
+    async with httpx.AsyncClient() as client:
+        resp = await client.post(url, json=payload, headers=headers, timeout=30)
+    data = resp.json()
+    if data.get("type") == "consentGranted":
+        return data["consentToken"]
+    else:
+        raise MissingConsentToken
+
+
+async def get_consent_verification_url(dsi: str, sub: str) -> Optional[str]:
+    """
+    Start consent flow. Request consent with a Consent Request Token,
+    try to get a verification URL back from Consent Portal.
+    :param dsi: Data Source Identifier
+    :param sub: Subject of consent
+    :return: URL to verify a consent in Consent Portal
+    """
+    url = f"{conf.CONSENT_PORTAL_URL}/Consent/RequestConsents"
+    cr_token = create_consent_request_token(sub)
+    headers = {"X-Consent-Request-Token": cr_token}
+    payload = {"consentRequests": [{"dataSource": dsi, "required": True}]}
+    async with httpx.AsyncClient() as client:
+        resp = await client.post(url, json=payload, headers=headers, timeout=30)
+    data = resp.json()
+    if data.get("type") == "requestUserConsent":
+        return data["requestUrl"]
+    return None
+
+
+@router.post("/data-product-consent/{data_product:path}")
+async def fetch_data_product(
+    data_product: str,
+    request: Request,
+    source=Query(),
+    id_token: Optional[str] = Cookie(default=None),
+):
+    """
+    A proxy from frontend to Product Gateway for data products that require consent.
+    First it tries to find consent token in a cookie and if not found it
+    tries to fetch a new token from Consent Portal. If it's an initial request,
+    then a user must grant the consent in Consent Portal UI.
+    """
+    # Validate ID Token to find a user ID that we use as a "sub" for consent request
+    sub = await parse_token(id_token)
+
+    dsi = make_dsi_uri(data_product)
+
+    try:
+        # Try to fetch a token from Consent Portal
+        # IMPORTANT! Here we request consent token every time.
+        # In real applications you might want to save the token in database
+        # or a cookie and reuse it to avoid making extra API calls
+        consent_token = await request_consent_token(dsi, sub)
```

Comment:
I'm a bit wondering if it'd look nicer to pass in the data_product and source to the function and let the internals of the function convert them to a DSI.